pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -2813,21 +2824,30 @@ public void removedConsumer(Consumer consumer) {
 
     @Override
     public void removedProducer(Producer producer) {
-        safelyRemoveProducer(producer);
+        safelyRemoveProducer(producer, false);
     }
 
-    private void safelyRemoveProducer(Producer producer) {
+    private void safelyRemoveProducer(Producer producer, boolean keepTombstoneForProducer) {
         long producerId = producer.getProducerId();
         if (log.isDebugEnabled()) {
             log.debug("[{}] Removed producer: producerId={}, producer={}", remoteAddress, producerId, producer);
         }
         CompletableFuture<Producer> future = producers.get(producerId);
         if (future != null) {
-            future.whenComplete((producer2, exception) -> {
+            future.whenCompleteAsync((producer2, exception) -> {
                     if (exception != null || producer2 == producer) {
                         producers.remove(producerId, future);
+                        if (keepTombstoneForProducer
+                                && isActive
+                                && getRemoteEndpointProtocolVersion() >= v5.getValue()) {
```

Comment:
Add a comment that explains the reason for this condition about protocol version `v5`

Reply:
My refactored solution should make this a bit clearer. We only need to keep track of the producers that we close via the `CloseProducer` command.

Reply:
Yes, it's easy to read it now. Thanks for addressing my comments.